### PR TITLE
Makes erlang cookie configurable

### PR DESCRIPTION
**Pull request by GoelDeepak**
## High-level description

Erlang cookie serves as a membership token in a distributed system. Having it configurable would allow different DC/OS clusters to co-exist without interfering each other. The requirement is that it should be unique and persistent during the lifecycle for a given DC/OS cluster. The idea is to use `cluster_id` as it satisfies both the requirements.

## Corresponding DC/OS tickets (obligatory)

These DC/OS JIRA ticket(s) must be updated (ideally closed) in the moment this PR lands:

  - [DCOS_OSS-4620](https://jira.mesosphere.com/browse/DCOS_OSS-4620) Make cluster identity configurable in dcos-net

## Checklist for all PRs

  - [x] Added a comprehensible changelog entry to `CHANGES.md` or explain why this is not a user-facing change:
  - [ ] Included a test which will fail if code is reverted but test is not. If there is no test please explain here: Before going with the tests, I want to get approval from the reviewers about the approach. Once I get the approval I'll add the integration test.
  - [ ] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [ ] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]
___
**PLEASE FILL IN THE TEMPLATE ABOVE** / **DO NOT REMOVE ANY SECTIONS ABOVE THIS LINE**

## Instructions and review process

**What is the review process and when will my changes land?**

All PRs require 2 approvals using GitHub's [pull request reviews](https://help.github.com/articles/about-pull-request-reviews/).

Reviewers should be:
* Developers who understand the code being modified.
* Developers responsible for code that interacts with or depends on the code being modified.

It is best to proactively ask for 2 reviews by @mentioning the candidate reviewers in the PR comments area. The responsibility is on the developer submitting the PR to follow-up with reviewers and make sure a PR is reviewed in a timely manner. Once a PR has **2 ship-it's**, **no red reviews**, and **all tests are green** it will be included in the [next train](https://github.com/dcos/dcos/blob/master/contributing.md).

### CODE REVIEW

File: gen/calc.py [hunk 1]
```diff
@@ -381,6 +382,12 @@ def validate_dns_search(dns_search):
     assert len(dns_search.split()) <= 6, "Must contain no more than 6 domains"
 
 
+def calculate_dcos_net_cluster_name(dcos_net_cluster_identity, cluster_name):
+    return 'minuteman' if dcos_net_cluster_identity == 'false' \
+           else cluster_name[:255] if len(cluster_name) > 250  \
+           else '-'.join([cluster_name, ''.join(random.choice(string.ascii_lowercase) for _ in range(5))])
```
Comment (urbanserj) [inline gen/calc.py]: Will it generate a new random cookie on each upgrade?

Comment (GoelDeepak) [reply gen/calc.py]: Excellent catch! I guess that is the case. Let me see what is the best way to fix it

File: packages/dcos-net/extra/dcos-net-env [hunk 2]
```diff
@@ -63,6 +64,18 @@ EOM
...
 
+function cookie() {
+    FILE="/var/lib/dcos/cluster-id"
+    if [ "${DCOS_NET_CLUSTER_IDENTITY}" == "false" ]; then
+        echo 'minuteman'
+    elif [ ! -f ${FILE} ]; then
+        echo "ERROR: required file ${FILE} does not exists" >&2
+        exit 1
+    else
+        echo "$(cat ${FILE})"
```
Comment (urbanserj) [inline packages/dcos-net/extra/dcos-net-env]: nit:

```
-        echo "$(cat ${FILE})"
+        cat ${FILE}
```

File: packages/bootstrap/extra/dcos_internal_utils/cli.py [hunk 3]
```diff
@@ -73,17 +73,22 @@ def dcos_adminrouter(b, opts):
...
 
 @check_root
 def dcos_telegraf_agent(b, opts):
-    b.cluster_id('/var/lib/dcos/cluster-id', readonly=True)
+    b.cluster_id(readonly=True)
+
+
+@check_root
+def dcos_net(b, opts):
+    b.cluster_id()
```
Comment (mhrabovcin) [inline packages/bootstrap/extra/dcos_internal_utils/cli.py]: Is this going to be executed same way on master and agent nodes? If yes, do we need to split bootstrap into a 2 separate code paths? See `dcos_telegraf_master` and `dcos_telegraf_agent` above which does `b.cluster_id(readonly=True)` on agent.

**Comment by mhrabovcin** [comment]
@GoelDeepak is the cookie value supposed to be treated as a secret? The `cluster-id` is a public interface and it is not meant to be used for any security purposes.

**Comment by mhrabovcin** [comment]
@GoelDeepak are you planning to add any integration and or unit test for the new configuration option?

**Comment by GoelDeepak** [comment]
> @GoelDeepak is the cookie value supposed to be treated as a secret? The `cluster-id` is a public interface and it is not meant to be used for any security purposes.

No, it is not supposed to be a secret

**Comment by adamtheturtle** [comment]
> @GoelDeepak are you planning to add any integration and or unit test for the new configuration option?

Thank you @mhrabovcin for pointing this out.
I'd like to go further and point out that there is a Checklist for all PRs which includes:

>  Included a test which will fail if code is reverted but test is not. If there is no test please explain here:

That is neither checked nor explained. As a reviewer, I look at this checklist to see if there is a valid justification for no tests.
Please fill in this check box and the others.

As I am someone who has written tests for some configuration values, I am happy to help if desired.

File: packages/bootstrap/extra/dcos_internal_utils/cli.py (hunk 3 above)
Comment (GoelDeepak) [reply packages/bootstrap/extra/dcos_internal_utils/cli.py]: Yeah, either `master` or `agent` can generate the ID and rest of the nodes would then use it

**Comment by GoelDeepak** [comment]
> @GoelDeepak are you planning to add any integration and or unit test for the new configuration option?

I wanted to run it through the first round of review to make sure it makes sense. I'll add the tests

**Comment by GoelDeepak** [comment]
> > @GoelDeepak are you planning to add any integration and or unit test for the new configuration option?
> 
> Thank you @mhrabovcin for pointing this out.
> I'd like to go further and point out that there is a Checklist for all PRs which includes:
> 
> > Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
> 
> That is neither checked nor explained. As a reviewer, I look at this checklist to see if there is a valid justification for no tests.
> Please fill in this check box and the others.
> 
> As I am someone who has written tests for some configuration values, I am happy to help if desired.

I am aware of the checklist but before going ahead with the tests I wanted to review the approach.

**Comment by GoelDeepak** [comment]
> > > @GoelDeepak are you planning to add any integration and or unit test for the new configuration option?
> > 
> > 
> > Thank you @mhrabovcin for pointing this out.
> > I'd like to go further and point out that there is a Checklist for all PRs which includes:
> > > Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
> > 
> > 
> > That is neither checked nor explained. As a reviewer, I look at this checklist to see if there is a valid justification for no tests.
> > Please fill in this check box and the others.
> > As I am someone who has written tests for some configuration values, I am happy to help if desired.
> 
> I am aware of the checklist but before going ahead with the tests I wanted to review the approach.

File: packages/bootstrap/extra/dcos_internal_utils/cli.py (hunk 3 above)
Comment (mhrabovcin) [reply packages/bootstrap/extra/dcos_internal_utils/cli.py]: I'd split bootstrap function to `dcos_net_master` and `dcos_net_agent` and use the `b.cluster_id(readonly=True)` for agent bootstrap.

**Comment by GoelDeepak** [comment]
@mhrabovcin @adamtheturtle would you mind if we follow the integration test in a separate PR? I realized that a proper test would involve launching a DC/OS cluster with a non-default configuration. Something that we currently support only in DC/OS Enterprise through test-e2e. It will take me some time to understand test-e2e and create a meaningful test. I have filed a jira ticket https://jira.mesosphere.com/browse/DCOS_OSS-4668 to track that effort.

Meanwhile, I have manually verified the changes through manual testing

**Comment by adamtheturtle** [comment]
> @mhrabovcin @adamtheturtle would you mind if we follow the integration test in a separate PR?

@GoelDeepak For what benefit would this be merged without a test?

**Comment by GoelDeepak** [comment]
> > @mhrabovcin @adamtheturtle would you mind if we follow the integration test in a separate PR?
> 
> @GoelDeepak For what benefit would this be merged without a test?

As I mentioned, the functionality has been tested manually. What is the harm in splitting the functionality and the test in two separate PRs? Further, even if I write the integration test it would only be applicable to DC/OS enterprise. DC/OS open would still remain without an integration test.

**Comment by adamtheturtle** [comment]
> What is the harm in splitting the functionality and the test in two separate PRs?

For some time we have even more code in DC/OS without regression tests.

We have a standard practice of adding tests with new functionality in order to avoid this downside.
That practice has exceptions, in particular when we need functionality urgently, before tests can be written.
I was asking "Why have an exception?".

There are multiple layers of tests that can be applied here, at the unit level and at the E2E level.

> even if I write the integration test it would only be applicable to DC/OS enterprise. DC/OS open would still remain without an integration test.

Is that because of https://jira.mesosphere.com/browse/DCOS_OSS-2032?

**Comment by GoelDeepak** [comment]
> I was asking "Why have an exception?".
Yes, there is a customer waiting for this functionality

> There are multiple layers of tests that can be applied here, at the unit level and at the E2E level.
At unit level, the config validator already ensures that correct config is being passed. We need one E2E test.

> Is that because of https://jira.mesosphere.com/browse/DCOS_OSS-2032?
Yes

**Comment by GoelDeepak** [comment]
I had a chat with @adamtheturtle and we agreed that we will have a test that would only be executed in DC/OS EE because of lack of e2e in DC/OS open. The test would still appear in DC/OS open but would not get executed and this is acceptable for now

**Comment by adamtheturtle** [comment]
> At unit level, the config validator already ensures that correct config is being passed. We need one E2E test.

See https://github.com/dcos/dcos/blob/master/gen/tests/test_adminrouter_tls_conf.py for the kinds of tests I was referring to.

That might include:

```
test_must_be_true_or_false (error shown if `foo` is given)
test_default (checks that the default config value is `false`)
test_config (checks that the generated config includes the given value)
```

File: packages/bootstrap/extra/dcos_internal_utils/cli.py (hunk 3 above)
Comment (GoelDeepak) [reply packages/bootstrap/extra/dcos_internal_utils/cli.py]: Done

File: packages/dcos-integration-test/extra/test_networking.py [hunk 4]
```diff
@@ -651,3 +651,17 @@ def enum2str(value):
 
 def net2str(value, ipv6):
     return enum2str(value) if not ipv6 else 'ipv6'
+
+
+def test_dcos_net_cluster_identity(dcos_api_session):
```
Comment (mhrabovcin) [inline packages/dcos-integration-test/extra/test_networking.py]: 💯

**Comment by orsenthil** [comment]
(the PR was already reviewed, and was rebased to include the fix for the flaky test)

**Pull request merged**